L1Trigger/TrackFindingTracklet/src/KFin.cc

```diff
@@ -0,0 +1,418 @@
+          if (ttStubRef.isNull())
+            continue;
+          // parse residuals from tt::Frame and take r and layerId from tt::TTStubRef
+          const bool barrel = setup_->barrel(ttStubRef);
+          const int layerIdTracklet = channelAssignment_->trackletLayerId(ttStubRef);
+          const double basePhi = barrel ? settings_->kphi1() : settings_->kphi(layerIdTracklet);
+          const double baseRZ = barrel ? settings_->kz(layerIdTracklet) : settings_->kz();
+          const int widthRZ = barrel ? setup_->tbWidthZ() : setup_->tbWidthR();
+          TTBV hw(frameStub.second);
+          const TTBV hwRZ(hw, widthRZ, 0, true);
+          hw >>= widthRZ;
+          const TTBV hwPhi(hw, setup_->tbWidthPhi(), 0, true);
+          hw >>= setup_->tbWidthPhi();
+          const double r = digi(setup_->stubR(hw, ttStubRef) - dataFormats_->chosenRofPhi(), baseUr_);
+          double phi = hwPhi.val(basePhi);
+          if (basePhi > baseUphi_)
+            phi += baseUphi_ / 2.;
+          const double z = digi(hwRZ.val(baseRZ) * (barrel ? 1. : -cot), baseUz_);
+          // determine module type
+          double rz = r + digi(dataFormats_->chosenRofPhi(), baseUr_);
+          if (barrel)
+            rz = rz * cot + z0 + z;
+          const int indexLayerId = setup_->indexLayerId(ttStubRef);
+          const double limit = barrel ? setup_->tiltedLayerLimitZ(indexLayerId) : setup_->psDiskLimitR(indexLayerId);
+          const bool psTilt = barrel ? (abs(rz) < limit) : setup_->psModule(ttStubRef);
+          stubs_.emplace_back(ttStubRef, layerId, r, phi, z, psTilt);
+          stubs.push_back(&stubs_.back());
+        }
+        // create fake seed stubs
+        for (int layerId : channelAssignment_->seedingLayers(channel)) {
+          const vector<TTStubRef>& ttStubRefs = ttTrackRef->getStubRefs();
+          auto sameLayer = [this, layerId](const TTStubRef& ttStubRef) {
+            return setup_->layerId(ttStubRef) == layerId;
+          };
+          const TTStubRef& ttStubRef = *find_if(ttStubRefs.begin(), ttStubRefs.end(), sameLayer);
+          const bool barrel = setup_->barrel(ttStubRef);
+          double r;
+          if (barrel)
+            r = digi(setup_->hybridLayerR(layerId - setup_->offsetLayerId()) - dataFormats_->chosenRofPhi(), baseUr_);
+          else {
+            r = (z0 +
+                 digi(setup_->hybridDiskZ(layerId - setup_->offsetLayerId() - setup_->offsetLayerDisks()), baseUzT_)) *
+                digi(1. / digi(abs(cot), baseCot), baseInvCot_);
+            r = digi(r - digi(dataFormats_->chosenRofPhi(), baseUr_), baseUr_);
+          }
+          static constexpr double phi = 0.;
+          static constexpr double z = 0.;
+          double rz = r;
+          if (barrel)
+            rz = abs(digi(digi(setup_->hybridLayerR(layerId - setup_->offsetLayerId()), baseUr_) * cot + z0, baseUz_));
+          const int indexLayerId = setup_->indexLayerId(ttStubRef);
+          const double limit = barrel
```

Comment:
Why does expression for "limit" in endcap differ at L142 & L170 in that one subtracts chosenR and the other does not? In any case, "limit" in the endcap is never used.

Reply:
I rewrote that a little